common/goMetricsExporter/go-metrics-m3.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package goMetricsExporter
+
+import (
+	gometrics "github.com/rcrowley/go-metrics"
+	m "github.com/uber/cherami-server/common/metrics"
+	"time"
+)
+
+const (
+	delayBetweenExports = time.Second * 10
+)
+
+// GoMetricsExporter allows an rcrowley/go-metrics registry to be
+// periodically exported to the Cherami metrics system
+type GoMetricsExporter struct {
+	m     m.Client
+	scope int
+
+	// metricNameToMetric maps a metric name to the metric symbol,
+	// e.g. "request-rate" -> OutputhostCGKafkaRequestRate
+	metricNameToMetric  map[string]int
+	closeCh             chan struct{}
+	metricNameToLastVal map[string]int64
+	registry            gometrics.Registry // Registry to be exported
+}
+
+// NewGoMetricsExporter starts the exporter loop to export go-metrics to the Cherami
+// metrics system and returns the go-metrics registry that should be used
+func NewGoMetricsExporter(
+	m m.Client,
+	scope int,
+	metricNameToMetric map[string]int,
+) (*GoMetricsExporter, gometrics.Registry) {
+	// Make a new registy and give it to the caller. This allows the exporters to operate independently
+	registry := gometrics.NewRegistry()
+	r := &GoMetricsExporter{
+		m:                   m,
+		scope:               scope,
+		registry:            registry,
+		metricNameToMetric:  metricNameToMetric,
+		closeCh:             make(chan struct{}, 0),
+		metricNameToLastVal: make(map[string]int64),
+	}
+	return r, registry
+}
+
+// Stop stops the exporter
+func (r *GoMetricsExporter) Stop() {
+	close(r.closeCh)
+}
+
+// Run runs the exporter. It blocks until Stop() is called, so it should probably be run in a go-routine
+func (r *GoMetricsExporter) Run() {
```

Comment:
having the caller do a "go Run()" seems to unnecessarily expose internals. I would (again) suggesting having a "Start()" that does a "go r.run()" internally. That goes well will the "Stop()" you have.